packages/flutter/lib/src/cupertino/segmented_control.dart

```diff
@@ -0,0 +1,568 @@
+      final Offset boxOffset = new Offset(x, (maxHeight - boxSize.height) / 2.0);
+      return boxOffset;
+    }
+
+    final double childWidth = overallSize.width / length;
+
+    final BoxConstraints childConstraints =
+      new BoxConstraints.tightFor(width: childWidth, height: maxHeight);
+
+    child = firstChild;
+    while (child != null) {
+      final _SegmentedControlContainerBoxParentData childParentData = child.parentData;
+      child.layout(childConstraints, parentUsesSize: true);
+      child = childParentData.nextSibling;
+    }
+
+    Offset childOffset = Offset.zero;
+
+    const double left = 0.0;
+    final double right = overallSize.width;
+
+    switch (textDirection) {
+      case TextDirection.rtl:
+        double start = right;
+        child = firstChild;
+        while (child != null) {
+          final _SegmentedControlContainerBoxParentData childParentData = child.parentData;
+          childOffset = centerLayout(Size(childWidth, maxHeight), start - childWidth);
+          childParentData.offset = childOffset;
+          final Rect childRect = Rect.fromLTWH(start - childWidth, 0.0, childWidth, maxHeight);
+          RRect roundedChildRect;
+          if (child == firstChild) {
+            roundedChildRect = RRect.fromRectAndCorners(childRect, topLeft: const Radius.circular(3.0),
+                bottomLeft: const Radius.circular(3.0));
+          } else if (child == lastChild) {
+            roundedChildRect = RRect.fromRectAndCorners(childRect, topRight: const Radius.circular(3.0),
+                bottomRight: const Radius.circular(3.0));
+          } else {
+            roundedChildRect = RRect.fromRectAndCorners(childRect);
+          }
+          _rects.add(roundedChildRect);
+          start -= childWidth;
+          child = childParentData.nextSibling;
+        }
+        break;
+      case TextDirection.ltr:
+        double start = left;
+        child = firstChild;
+        while (child != null) {
+          final _SegmentedControlContainerBoxParentData childParentData = child.parentData;
+          childOffset = centerLayout(Size(childWidth, maxHeight),
+              start - _boxSize(child).width + childWidth);
+          childParentData.offset = childOffset;
+          final Rect childRect = Rect.fromLTWH(start, 0.0, childWidth, maxHeight);
+          RRect rChildRect;
+          if (child == firstChild) {
+            rChildRect = RRect.fromRectAndCorners(childRect, topLeft: const Radius.circular(3.0),
+                bottomLeft: const Radius.circular(3.0));
+          } else if (child == lastChild) {
+            rChildRect = RRect.fromRectAndCorners(childRect, topRight: const Radius.circular(3.0),
+                bottomRight: const Radius.circular(3.0));
+          } else {
+            rChildRect = RRect.fromRectAndCorners(childRect);
+          }
+          _rects.add(rChildRect);
+          start += childWidth;
+          child = childParentData.nextSibling;
+        }
+        break;
```

Comment:
to avoid code duplication here you could factor out this logic, having it always do the list from left to right, but taking as arguments three values: a callback that returns the next child (the `nextChild` and `previousChild` methods exist to do this already), and the left child and the right child (passing in `firstChild` or `lastChild` accordingly).